File: Shared/ValidationTableUtilities.vb
'*********************************************************************************************
' ST-Sim: A SyncroSim Module for the ST-Sim State-and-Transition Model.
'
' Copyright © 2007-2017 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.
'
'*********************************************************************************************

Imports SyncroSim.Core
Imports SyncroSim.StochasticTime
Imports System.Globalization

Module ValidationTableUtilities

    ''' <summary>
    ''' Creates an age validation table
    ''' </summary>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public Function CreateAgeValidationTable(ByVal project As Project) As ValidationTable

        Dim dt As New DataTable(AGE_VALIDATION_TABLE_NAME)
        dt.Locale = CultureInfo.InvariantCulture

        dt.Columns.Add(New DataColumn(VALUE_MEMBER_COLUMN_NAME, GetType(Long)))
        dt.Columns.Add(New DataColumn(DISPLAY_MEMBER_COLUMN_NAME, GetType(String)))

        Dim e As IEnumerable(Of AgeDescriptor) = GetAgeGroupDescriptors(project)

        If (e Is Nothing) Then
            e = GetAgeTypeDescriptors(project)
        End If

        If (e IsNot Nothing) Then

            For Each d As AgeDescriptor In e

                Dim Value As Long = CLng(d.MinimumAge)
                Dim Display As String = Nothing

                If (d.MaximumAge.HasValue) Then

                    If (d.MaximumAge.Value = d.MinimumAge) Then
                        Display = String.Format(CultureInfo.InvariantCulture, "{0}", d.MinimumAge)
                    Else
                        Display = String.Format(CultureInfo.InvariantCulture, "{0}-{1}", d.MinimumAge, d.MaximumAge.Value)
                    End If

                Else
                    Display = String.Format(CultureInfo.InvariantCulture, "{0}+", d.MinimumAge)
                End If

                dt.Rows.Add({Value, Display})

            Next

        End If

        Return New ValidationTable(
            dt,
            VALUE_MEMBER_COLUMN_NAME,
            DISPLAY_MEMBER_COLUMN_NAME,
            SortOrder.None)

    End Function

End Module

File: Shared/ValidationTableUtilities.cs
// ST-Sim: A SyncroSim Module for the ST-Sim State-and-Transition Model.
// Copyright © 2007-2018 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.

using System;
using System.Data;
using SyncroSim.Core;
using System.Globalization;
using System.Collections.Generic;

namespace SyncroSim.STSim
{
    internal static class ValidationTableUtilities
    {
        /// <summary>
        /// Creates an age validation table
        /// </summary>
        /// <returns></returns>
        /// <remarks></remarks>
        public static ValidationTable CreateAgeValidationTable(Project project)
        {
            DataTable dt = new DataTable(Strings.AGE_VALIDATION_TABLE_NAME);
            dt.Locale = CultureInfo.InvariantCulture;

            dt.Columns.Add(new DataColumn(Strings.VALUE_MEMBER_COLUMN_NAME, typeof(long)));
            dt.Columns.Add(new DataColumn(Strings.DISPLAY_MEMBER_COLUMN_NAME, typeof(string)));

            IEnumerable<AgeDescriptor> e = AgeUtilities.GetAgeGroupDescriptors(project);

            if (e == null)
            {
                e = AgeUtilities.GetAgeTypeDescriptors(project);
            }

            if (e != null)
            {
                foreach (AgeDescriptor d in e)
                {
                    long Value = Convert.ToInt64(d.MinimumAge);
                    string Display = null;

                    if (d.MaximumAge.HasValue)
                    {
                        if (d.MaximumAge.Value == d.MinimumAge)
                        {
                            Display = string.Format(CultureInfo.InvariantCulture, "{0}", d.MinimumAge);
                        }
                        else
                        {
                            Display = string.Format(CultureInfo.InvariantCulture, "{0}-{1}", d.MinimumAge, d.MaximumAge.Value);
                        }
                    }
                    else
                    {
                        Display = string.Format(CultureInfo.InvariantCulture, "{0}+", d.MinimumAge);
                    }

                    dt.Rows.Add(new object[] {Value, Display});
                }
            }

            return new ValidationTable(dt, Strings.VALUE_MEMBER_COLUMN_NAME, Strings.DISPLAY_MEMBER_COLUMN_NAME, SortOrder.None);
        }
    }
}
